### PR TITLE
feature: custom templates / download default template

### PR DESCRIPTION
**Pull request by KiraGreifeneder**
Fixes issue where the template file wasn't downloaded during installation from the community plugins tab. Solved by downloading it from the repo on note creation when necessary.

Adds the ability for users to define a custom template.

- Added settings field that allows user to specify a path to a custom template
- Added functions in xoppActions.ts to get user template path if it is set or the default template
  - Added a util function in downloader.ts

This is my first PR on a repo like this so let me know if anything is weird or changes need to be made 😅 greetings from Austria

### CODE REVIEW

File: src/downloader.ts
Comment (jonjampen) [inline src/downloader.ts:1]: According to the Obsidian Developer Policies, a plugin is not allowed to download files over the network:

> Themes may not load assets from the network. [Obsidian Developer Policies](https://docs.obsidian.md/Developer+policies#Not+allowed)

I think we have to include the template in the code (as the file is quite small, that should not be an issue). The steps would probably be:

1. Convert .xopp template to base64 and store that as a string in a js file
2. Import the base64 string from the file into a `createTemplate()` function
3. This function converts the base64 to an arrayBuffer and then creates a file in the file system using the [createBinary method](https://docs.obsidian.md/Reference/TypeScript+API/Vault/createBinary#Vault.createBinary()+method)

Or do you have another idea?

Would you be able to implement those changes? Thank you for your contribution!

**Review by jonjampen — CHANGES_REQUESTED** [review]
Thank you for your contribution!

Please use indentation with spaces (4 spaces per indentation level) instead of tabs for consistency.

**Comment by jonjampen** [comment]
Thank you for developing the custom template feature, that is great!

File: src/downloader.ts
Comment (KiraGreifeneder) [reply src/downloader.ts:1]: Oh wow thank you so much for looking into that, I didn't realize that could be an issue. Keeping it in the code is a much better idea, I'll work on it sometime this week, along with fixing the indentation.

Thanks for taking the time!

Comment (jonjampen) [reply src/downloader.ts:1]: That's awesome, thank you!

**Comment by KiraGreifeneder** [comment]
Hi, I implemented your requested changes.

Btw, I later noticed that the Obsidian guidelines only say that _themes_ can't download assets, so I think plugins are allowed to.
I think your approach is still better though, so I stuck to your recommendation.

- stored the template as a base64 string in template.ts
- replaced downloadTemplate() with a createTemplate() function that converts the string to an ArrayBuffer and creates the template file from it

**Review by jonjampen — APPROVED** [review]
Thank you so much for taking the time to improve the plugin!
Your changes look great, I will merge it and create a new patch release.

**Comment by jonjampen** [comment]
> Btw, I later noticed that the Obsidian guidelines only say that _themes_ can't download assets, so I think plugins are allowed to. I think your approach is still better though, so I stuck to your recommendation.

You are right! But I agree, that it is still better to keep the plugin running offline 👍.

**Comment by KiraGreifeneder** [comment]
thanks so much for your patience and reviews!

**Pull request merged**